storage/src/journal/variable.rs

```diff
@@ -1,2223 +1,1443 @@
-//!
-//! # Example
-//!
-//! ```rust
-//! use commonware_runtime::{Spawner, Runner, deterministic, buffer::PoolRef};
-//! use commonware_storage::journal::variable::{Journal, Config};
-//! use commonware_utils::NZUsize;
-//!
-//! let executor = deterministic::Runner::default();
-//! executor.start(|context| async move {
-//!     // Create a journal
-//!     let mut journal = Journal::init(context, Config{
-//!         partition: "partition".to_string(),
-//!         compression: None,
-//!         codec_config: (),
-//!         buffer_pool: PoolRef::new(NZUsize!(1024), NZUsize!(10)),
-//!         write_buffer: NZUsize!(1024 * 1024),
-//!     }).await.unwrap();
-//!
-//!     // Append data to the journal
-//!     journal.append(1, 128).await.unwrap();
-//!
-//!     // Close the journal
-//!     journal.close().await.unwrap();
-//! });
-//! ```
+//! This journal enforces section fullness: all non-final sections are full and synced.
+//! On init, only the last section needs to be replayed to determine the exact size.
 
-use super::Error;
-use bytes::BufMut;
-use commonware_codec::Codec;
-use commonware_runtime::{
-    buffer::{Append, PoolRef, Read},
-    Blob, Error as RError, Metrics, Storage,
+use super::Journal as JournalTrait;
+use crate::{
+    journal::{fixed, Error},
+    multijournal,
 };
-use commonware_utils::hex;
-use futures::stream::{self, Stream, StreamExt};
-use prometheus_client::metrics::{counter::Counter, gauge::Gauge};
+use commonware_codec::Codec;
+use commonware_runtime::{buffer::PoolRef, Metrics, Storage};
+use commonware_utils::NZUsize;
+use futures::{stream, Stream, StreamExt as _};
 use std::{
-    collections::{btree_map::Entry, BTreeMap},
-    io::Cursor,
-    marker::PhantomData,
-    num::NonZeroUsize,
+    num::{NonZeroU64, NonZeroUsize},
+    pin::Pin,
 };
-use tracing::{debug, trace, warn};
-use zstd::{bulk::compress, decode_all};
+use tracing::info;
+
+const REPLAY_BUFFER_SIZE: NonZeroUsize = NZUsize!(1024);
+
+/// Calculate the section number for a given position.
+///
+/// # Arguments
+///
+/// * `position` - The absolute position in the journal
+/// * `items_per_section` - The number of items stored in each section
+///
+/// # Returns
+///
+/// The section number where the item at `position` should be stored.
+///
+/// # Examples
+///
+/// ```ignore
+/// // With 10 items per section:
+/// assert_eq!(position_to_section(0, 10), 0);   // position 0 -> section 0
+/// assert_eq!(position_to_section(9, 10), 0);   // position 9 -> section 0
+/// assert_eq!(position_to_section(10, 10), 1);  // position 10 -> section 1
+/// assert_eq!(position_to_section(25, 10), 2);  // position 25 -> section 2
+/// assert_eq!(position_to_section(30, 10), 3);  // position 30 -> section 3
+/// ```
+const fn position_to_section(position: u64, items_per_section: u64) -> u64 {
+    position / items_per_section
+}
 
-/// Configuration for `Journal` storage.
+/// Configuration for a variable-length journal.
```

Comment:
"variable-length journal" is sort of ambiguous, I prefer to say "journal for variable-length items" or similar.